#Adam Kritz Crate and Barrel Project

Welcome to my second project for data mining, the crate and barrel project. Today I will be classifying products into their proper categories for the Otto Group. Now in real life I may be a far left anarcho-socialist, but for today only I will bend the knee to large coporations, and do the bidding of the Otto Group. 

Contents:

1. Get those files in
2. Explore the data
3. Split the data
4. Bagging and Pasting
5. Random Subspace and Random Patches
6. XGBoost
7. Neural Network
8. Final Training
9. Conclusion


#Read the files

To start, I will upload the files and read them in. The following commands will only work properly if you have the same files, the training and testing data from the Kaggle website for the Otto Group project. The files must also be the same name as mine, "kaggletrain.csv" and "kaggletest.csv". Lastly, they must be in the same place as my files, uploaded alongside the "sample_data". 

In [1]:
!ls

kaggletest.csv	kaggletrain.csv  sample_data


In [2]:
import pandas as pd
kaggletrain = pd.read_csv('kaggletrain.csv', index_col = 'id')
kaggletest = pd.read_csv('kaggletest.csv', index_col = 'id')

I went ahead and imported the ID as the index to save myself some trouble later.

#Explore the data

Now it is time to see what this data is all about. I already know from the directions that I will be classifying the data categorically. The features are all in numerical data, with counts representing the occurences of different events. I also know there are 93 features to look at and more than 200,000 products. Now to take a look at the data.

In [3]:
kaggletrain.head()

feat_1  feat_2  feat_3  feat_4  ...  feat_91  feat_92  feat_93   target
id                                  ...                                    
1        1       0       0       0  ...        0        0        0  Class_1
2        0       0       0       0  ...        0        0        0  Class_1
3        0       0       0       0  ...        0        0        0  Class_1
4        1       0       0       1  ...        0        0        0  Class_1
5        0       0       0       0  ...        0        0        0  Class_1

[5 rows x 94 columns]

In [4]:
len(kaggletrain)

61878

In [5]:
kaggletest.head()

feat_1  feat_2  feat_3  feat_4  ...  feat_90  feat_91  feat_92  feat_93
id                                  ...                                    
1        0       0       0       0  ...        0        0        0        0
2        2       2      14      16  ...        0        0        2        0
3        0       1      12       1  ...        0        0        0        1
4        0       0       0       1  ...        0        0        0        0
5        1       0       0       1  ...        0        9        0        0

[5 rows x 93 columns]

In [6]:
len(kaggletest)

144368

Already off the bat there is some interesting stuff to note. The training set is actually far smaller than the testing set, which I have not dealt with yet. This may affect how accurate certain classifiers are. On top of this, the test set does not actually have the true labels for the data in it. This was likely to prevent cheating in the competition. 

Now to take a look at the description of the data.

In [7]:
kaggletrain.describe(include="all")

feat_1        feat_2  ...       feat_93   target
count   61878.00000  61878.000000  ...  61878.000000    61878
unique          NaN           NaN  ...           NaN        9
top             NaN           NaN  ...           NaN  Class_2
freq            NaN           NaN  ...           NaN    16122
mean        0.38668      0.263066  ...      0.126135      NaN
std         1.52533      1.252073  ...      1.201720      NaN
min         0.00000      0.000000  ...      0.000000      NaN
25%         0.00000      0.000000  ...      0.000000      NaN
50%         0.00000      0.000000  ...      0.000000      NaN
75%         0.00000      0.000000  ...      0.000000      NaN
max        61.00000     51.000000  ...     87.000000      NaN

[11 rows x 94 columns]

In [8]:
kaggletest.describe(include="all")

feat_1         feat_2  ...        feat_92        feat_93
count  144368.000000  144368.000000  ...  144368.000000  144368.000000
mean        0.386201       0.263597  ...       0.388348       0.132675
std         1.468882       1.261908  ...       1.006935       1.302695
min         0.000000       0.000000  ...       0.000000       0.000000
25%         0.000000       0.000000  ...       0.000000       0.000000
50%         0.000000       0.000000  ...       0.000000       0.000000
75%         0.000000       0.000000  ...       0.000000       0.000000
max        64.000000      45.000000  ...      22.000000      91.000000

[8 rows x 93 columns]

It seems like for both data sets the most common value is 0, with some higher values mixed in every once in a while that bring up the mean. I will test this theory by looking at the value counts of one of the columns in the training set.

In [9]:
kaggletrain['feat_1'].value_counts()

0     51483
1      5906
2      1829
3       981
4       521
5       471
6       207
7       192
8        71
9        55
22       32
11       24
10       15
13       10
15        8
26        6
19        5
24        5
12        5
28        4
14        4
21        4
23        4
17        4
48        3
16        3
27        3
25        3
39        2
31        2
42        2
43        2
30        2
47        2
56        1
20        1
37        1
34        1
29        1
61        1
32        1
40        1
Name: feat_1, dtype: int64

It seems like my prediction was relatively correct, higher numbers of events occur far less commonly than lower numbers.

Lastly I will just double check that both data sets do not contain null values that could mess up classifiers. 

In [10]:
kaggletrain.isnull().values.any()

False

In [11]:
kaggletest.isnull().values.any()

False

Excellent!

#Split the Data

Instead of using the test data to check accuracy, I will just split the training data into new testing and training groups, as I cannot repeatedly submit to kaggle to check accuracy. In this case I will not control randomness so I can test my shorter classifiers on multiple random sets.

In [12]:
ktrainlabels = kaggletrain['target']
ktrainfeat = kaggletrain.drop('target', axis = 1)
from sklearn.model_selection import train_test_split

In [13]:
ktrainf, ktestf, ktrainl, ktestl = train_test_split(ktrainfeat, ktrainlabels, test_size = 0.2)

Everything seems to be in order, now onto training.

#Bagging and Pasting 

I will start by trying out Bagging and Pasting on both decision tree and KNN classifiers.

In [14]:
from sklearn.ensemble import BaggingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree
from sklearn.metrics import accuracy_score


clf = tree.DecisionTreeClassifier(criterion='entropy')
bagging_clf = BaggingClassifier(clf, n_estimators=20, max_samples=100, 
                                bootstrap=True, n_jobs=-1)
bagging_clf.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [15]:
clf = tree.DecisionTreeClassifier(criterion='entropy')
pasting_clf = BaggingClassifier(clf, n_estimators=20, max_samples=100, 
                                bootstrap=False, n_jobs=-1)
pasting_clf.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [16]:
knn = KNeighborsClassifier(n_neighbors=3)
bagging_knn = BaggingClassifier(knn, n_estimators=20, max_samples=100, 
                                bootstrap=True, n_jobs=-1)
bagging_knn.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [17]:
knn = tree.DecisionTreeClassifier(criterion='entropy')
pasting_knn = BaggingClassifier(knn, n_estimators=20, max_samples=100, 
                                bootstrap=False, n_jobs=-1)
pasting_knn.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

This is a pretty atrocious start after a couple of runs, I will come back to these algorithms later to fix them up if the others are also not performing well.

#Random Subspace and Random Patches

Next I will try random subspace and patches again on decision tree and knn classifiers.

In [18]:
clf = tree.DecisionTreeClassifier(criterion='entropy')
subspace_clf = BaggingClassifier(clf, n_estimators=50, max_features=7, 
                                bootstrap=True, n_jobs=-1)
subspace_clf.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [19]:
clf = tree.DecisionTreeClassifier(criterion='entropy')
patches_clf = BaggingClassifier(clf, n_estimators=100, max_features=7, 
                                 max_samples=100, bootstrap=False, n_jobs=-1)
patches_clf.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [20]:
knn = tree.DecisionTreeClassifier(criterion='entropy')
subspace_knn = BaggingClassifier(knn, n_estimators=50, max_features=7, 
                                bootstrap=True, n_jobs=-1)
subspace_knn.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

In [21]:
knn = tree.DecisionTreeClassifier(criterion='entropy')
patches_knn = BaggingClassifier(knn, n_estimators=100, max_features=7, 
                                 max_samples=100, bootstrap=False, n_jobs=-1)
patches_knn.fit(ktrainf, ktrainl)
predictions = bagging_clf.predict(ktestf)
accuracy_score(ktestl, predictions)

0.622737556561086

Again these classifiers are not performing the best, so I will come back to them later if the others fail.

#XGBoost

Time to bring out the big guns, I will try an XGBoost classifier.

In [22]:
from xgboost import XGBClassifier
params = {"n_estimators": 400, 'tree_method':'gpu_hist', 'predictor':'gpu_predictor'}
Boost = XGBClassifier(**params)

Boost.fit(ktrainf, ktrainl)
predictions = Boost.predict(ktestf)
accuracy_score(ktestl, predictions)

0.7954104718810601

Holy moly, 79% accuracy right off the back, that blows the other ones out of the water. Let me see if I can improve anymore. I should note I originally tried using a gridsearch for this, but it was far too slow, even with minimal parameters, so I changed my mind.

Changing the max depth. This seemed to help a lot.

In [23]:
params = {"n_estimators": 400, "max_depth": 12, 'tree_method':'gpu_hist', 'predictor':'gpu_predictor'}
Boost = XGBClassifier(**params)

Boost.fit(ktrainf, ktrainl)
predictions = Boost.predict(ktestf)
accuracy_score(ktestl, predictions)

0.8265190691661279


Adjusting the learning rate seemed beneficial in some cases, but not in others, so I just left it as the default.


In [24]:
params = {"n_estimators": 400, "max_depth": 12, 'learning_rate': 0.3, 'tree_method':'gpu_hist', 'predictor':'gpu_predictor'}
Boost = XGBClassifier(**params)

Boost.fit(ktrainf, ktrainl)
predictions = Boost.predict(ktestf)
accuracy_score(ktestl, predictions)

0.8257110536522301

Adjusting the subsample to prevent overfitting. This did not seem to help.

In [25]:
params = {"n_estimators": 400, "max_depth": 12, 'learning_rate': 0.3, 'subsample': 0.5, 'tree_method':'gpu_hist', 'predictor':'gpu_predictor'}
Boost = XGBClassifier(**params)

Boost.fit(ktrainf, ktrainl)
predictions = Boost.predict(ktestf)
accuracy_score(ktestl, predictions)

0.8253070458952811

I will go with the standard XGBoost with a max_depth of 12.

#Neural Network

After an impressive showing by XGBoost, I shall see how a neural network performs. 

Before creating a neural network, I need to do some work on the labels to get them in the right format.

In [55]:
ktrainl.value_counts()

Class_2    12924
Class_6    11308
Class_8     6770
Class_3     6357
Class_9     3978
Class_7     2276
Class_5     2170
Class_4     2166
Class_1     1553
Name: target, dtype: int64

In [66]:
k_trainl = ktrainl.replace({'Class_1': 0, 'Class_2': 1, 'Class_3': 2, 'Class_4': 3, 'Class_5':4,
                            'Class_6': 5, 'Class_7': 6, 'Class_8': 7, 'Class_9': 8})
k_testl = ktestl.replace({'Class_1': 0, 'Class_2': 1, 'Class_3': 2, 'Class_4': 3, 'Class_5':4,
                            'Class_6': 5, 'Class_7': 6, 'Class_8': 7, 'Class_9': 8})
k_trainl

id
52206    7
26746    3
7562     1
8110     1
25041    2
        ..
7161     1
13239    1
7887     1
47564    6
42964    5
Name: target, Length: 49502, dtype: int64

In [69]:
from keras.utils import to_categorical

ktrain_labels = to_categorical(k_trainl, num_classes=9)
ktest_labels = to_categorical(k_testl, num_classes=9)

ktrain_labels 

array([[0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [70]:
ktrain_labels[0]

array([0., 0., 0., 0., 0., 0., 0., 1., 0.], dtype=float32)

Now for the neural network.

In [72]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=100)

Epoch 1/15
496/496 [==============================] - 1s 2ms/step - loss: 0.8257 - accuracy: 0.7216
Epoch 2/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5768 - accuracy: 0.7857
Epoch 3/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5270 - accuracy: 0.7989
Epoch 4/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4893 - accuracy: 0.8129
Epoch 5/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4666 - accuracy: 0.8220
Epoch 6/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4522 - accuracy: 0.8283
Epoch 7/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4310 - accuracy: 0.8343
Epoch 8/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4098 - accuracy: 0.8430
Epoch 9/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4019 - accuracy: 0.8446
Epoch 10/15
496/496 [==============================] - 1s 2ms/step - loss: 0.3771 - accuracy: 0.8548

In [73]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 0.5974 - accuracy: 0.8046
Accuracy: 0.8046218752861023


This a pretty great start for the neural network, I will see if I can change up the hyper paramters to make it even better. 

For starters, I'll add another layer.

In [74]:
network = models.Sequential()
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(200, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=100)

Epoch 1/15
496/496 [==============================] - 2s 2ms/step - loss: 0.7718 - accuracy: 0.7226
Epoch 2/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5427 - accuracy: 0.7849
Epoch 3/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5002 - accuracy: 0.8006
Epoch 4/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4707 - accuracy: 0.8133
Epoch 5/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4513 - accuracy: 0.8197
Epoch 6/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4262 - accuracy: 0.8293
Epoch 7/15
496/496 [==============================] - 1s 2ms/step - loss: 0.4073 - accuracy: 0.8367
Epoch 8/15
496/496 [==============================] - 1s 3ms/step - loss: 0.3861 - accuracy: 0.8445
Epoch 9/15
496/496 [==============================] - 1s 2ms/step - loss: 0.3656 - accuracy: 0.8564
Epoch 10/15
496/496 [==============================] - 1s 2ms/step - loss: 0.3467 - accuracy: 0.8615

In [75]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 0.5494 - accuracy: 0.8087
Accuracy: 0.808661937713623


Not much better.

Now I will change up the batch size.

In [76]:
network = models.Sequential()
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(200, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=20)

Epoch 1/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.7032 - accuracy: 0.7402
Epoch 2/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.5424 - accuracy: 0.7922
Epoch 3/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.5158 - accuracy: 0.8008
Epoch 4/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4834 - accuracy: 0.8152
Epoch 5/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4719 - accuracy: 0.8192
Epoch 6/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4512 - accuracy: 0.8274
Epoch 7/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4455 - accuracy: 0.8317
Epoch 8/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4251 - accuracy: 0.8391
Epoch 9/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.4200 - accuracy: 0.8404
Epoch 10/15
2476/2476 [==============================] - 6s 2ms/step - loss: 0.406

In [77]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 0.6143 - accuracy: 0.8013
Accuracy: 0.8013089895248413


I am going the wrong direction on my accuracy.

How about changing some activation conditions?

In [78]:
network = models.Sequential()
network.add(layers.Dense(500, activation='swish'))
network.add(layers.Dense(200, activation='tanh'))
network.add(layers.Dense(9, activation='gelu'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=100)

Epoch 1/15
496/496 [==============================] - 2s 2ms/step - loss: 2.4249 - accuracy: 0.6415
Epoch 2/15
496/496 [==============================] - 1s 2ms/step - loss: 1.6428 - accuracy: 0.6985
Epoch 3/15
496/496 [==============================] - 1s 2ms/step - loss: 2.6786 - accuracy: 0.6619
Epoch 4/15
496/496 [==============================] - 1s 2ms/step - loss: 3.1706 - accuracy: 0.6291
Epoch 5/15
496/496 [==============================] - 1s 2ms/step - loss: 4.6459 - accuracy: 0.5683
Epoch 6/15
496/496 [==============================] - 1s 2ms/step - loss: 3.3419 - accuracy: 0.6358
Epoch 7/15
496/496 [==============================] - 1s 2ms/step - loss: 3.8837 - accuracy: 0.6060
Epoch 8/15
496/496 [==============================] - 1s 2ms/step - loss: 4.2186 - accuracy: 0.5879
Epoch 9/15
496/496 [==============================] - 1s 2ms/step - loss: 3.6830 - accuracy: 0.6508
Epoch 10/15
496/496 [==============================] - 1s 2ms/step - loss: 3.4966 - accuracy: 0.6502

In [79]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 5.3248 - accuracy: 0.6196
Accuracy: 0.6195862889289856


Never mind that sucked.

Let me try to make a few more changes.

In [80]:
network = models.Sequential()
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(200, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=50, batch_size=100)

Epoch 1/50
496/496 [==============================] - 2s 2ms/step - loss: 0.7901 - accuracy: 0.7162
Epoch 2/50
496/496 [==============================] - 1s 2ms/step - loss: 0.5449 - accuracy: 0.7849
Epoch 3/50
496/496 [==============================] - 1s 2ms/step - loss: 0.5033 - accuracy: 0.8033
Epoch 4/50
496/496 [==============================] - 1s 2ms/step - loss: 0.4758 - accuracy: 0.8112
Epoch 5/50
496/496 [==============================] - 1s 2ms/step - loss: 0.4541 - accuracy: 0.8198
Epoch 6/50
496/496 [==============================] - 1s 2ms/step - loss: 0.4238 - accuracy: 0.8314
Epoch 7/50
496/496 [==============================] - 1s 2ms/step - loss: 0.4110 - accuracy: 0.8349
Epoch 8/50
496/496 [==============================] - 1s 3ms/step - loss: 0.3847 - accuracy: 0.8471
Epoch 9/50
496/496 [==============================] - 1s 2ms/step - loss: 0.3721 - accuracy: 0.8494
Epoch 10/50
496/496 [==============================] - 1s 2ms/step - loss: 0.3540 - accuracy: 0.8557

In [81]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 1.0541 - accuracy: 0.7962
Accuracy: 0.7962185144424438


In [82]:
network = models.Sequential()
network.add(layers.Dense(500, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=50)

Epoch 1/15
991/991 [==============================] - 2s 2ms/step - loss: 0.8013 - accuracy: 0.7171
Epoch 2/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5924 - accuracy: 0.7744
Epoch 3/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5564 - accuracy: 0.7844
Epoch 4/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5428 - accuracy: 0.7898
Epoch 5/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5268 - accuracy: 0.7957
Epoch 6/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5149 - accuracy: 0.8003
Epoch 7/15
991/991 [==============================] - 2s 2ms/step - loss: 0.5007 - accuracy: 0.8048
Epoch 8/15
991/991 [==============================] - 2s 2ms/step - loss: 0.4932 - accuracy: 0.8099
Epoch 9/15
991/991 [==============================] - 2s 2ms/step - loss: 0.4821 - accuracy: 0.8129
Epoch 10/15
991/991 [==============================] - 2s 2ms/step - loss: 0.4730 - accuracy: 0.8156

In [83]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 0.5481 - accuracy: 0.7868
Accuracy: 0.7867646813392639


In [84]:
network = models.Sequential()
network.add(layers.Dense(20, activation='relu'))
network.add(layers.Dense(20, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainf, ktrain_labels, epochs=15, batch_size=100)

Epoch 1/15
496/496 [==============================] - 2s 2ms/step - loss: 1.4210 - accuracy: 0.5508
Epoch 2/15
496/496 [==============================] - 1s 2ms/step - loss: 0.7174 - accuracy: 0.7400
Epoch 3/15
496/496 [==============================] - 1s 2ms/step - loss: 0.6332 - accuracy: 0.7591
Epoch 4/15
496/496 [==============================] - 1s 2ms/step - loss: 0.6136 - accuracy: 0.7660
Epoch 5/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5942 - accuracy: 0.7738
Epoch 6/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5845 - accuracy: 0.7758
Epoch 7/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5735 - accuracy: 0.7821
Epoch 8/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5741 - accuracy: 0.7792
Epoch 9/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5675 - accuracy: 0.7839
Epoch 10/15
496/496 [==============================] - 1s 2ms/step - loss: 0.5558 - accuracy: 0.7877

In [85]:
kaggle_loss, kaggle_acc = network.evaluate(ktestf, ktest_labels)
print('Accuracy:', kaggle_acc)

387/387 [==============================] - 1s 2ms/step - loss: 0.5734 - accuracy: 0.7845
Accuracy: 0.7845022678375244


I will use all the information I have here to come up with the best hyperparameters for the final training. 

#Final Training

Since I get two submissions to Kaggle, I will try both XGBoost and a Neural Network, as they both performed well. I will use the best performing hyper parameters from the training and and apply those to the full training data before I officially submit it to Kaggle.

In [113]:
params = { "n_estimators": 400, 'max_depth': 12, 'tree_method':'gpu_hist', 'predictor':'gpu_predictor' }
Boost = XGBClassifier(**params)

Boost.fit(ktrainfeat, ktrainlabels)
Boostpredictions = Boost.predict(kaggletest)
Boostpredictions

array(['Class_4', 'Class_8', 'Class_6', ..., 'Class_2', 'Class_4',
       'Class_3'], dtype=object)

In [114]:
boost = pd.Series(Boostpredictions)
boost

0         Class_4
1         Class_8
2         Class_6
3         Class_2
4         Class_9
           ...   
144363    Class_1
144364    Class_3
144365    Class_2
144366    Class_4
144367    Class_3
Length: 144368, dtype: object

In [116]:
boost2 = boost.replace({'Class_1': 0, 'Class_2': 1, 'Class_3': 2, 'Class_4': 3, 'Class_5':4,
                                   'Class_6': 5, 'Class_7': 6, 'Class_8': 7, 'Class_9':8})
boost3 = to_categorical(boost2)
boost3[0:3]

array([[0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 1., 0., 0., 0.]], dtype=float32)

In [117]:
boost4= pd.DataFrame(boost3)
boost4.head()

0    1    2    3    4    5    6    7    8
0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0
2  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0
3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0

In [118]:
boost4.index += 1 
boost4.index.name = 'id'
boost4.columns=['Class_1','Class_2','Class_3','Class_4','Class_5','Class_6','Class_7','Class_8','Class_9']
boost4.head()

Class_1  Class_2  Class_3  Class_4  ...  Class_6  Class_7  Class_8  Class_9
id                                      ...                                    
1       0.0      0.0      0.0      1.0  ...      0.0      0.0      0.0      0.0
2       0.0      0.0      0.0      0.0  ...      0.0      0.0      1.0      0.0
3       0.0      0.0      0.0      0.0  ...      1.0      0.0      0.0      0.0
4       0.0      1.0      0.0      0.0  ...      0.0      0.0      0.0      0.0
5       0.0      0.0      0.0      0.0  ...      0.0      0.0      0.0      1.0

[5 rows x 9 columns]

In [86]:
ktl = ktrainlabels.replace({'Class_1': 0, 'Class_2': 1, 'Class_3': 2, 'Class_4': 3, 'Class_5':4,
                            'Class_6': 5, 'Class_7': 6, 'Class_8': 7, 'Class_9':8})
ktl = to_categorical(ktl)
ktl

array([[1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 0., ..., 0., 0., 1.]], dtype=float32)

In [88]:
network = models.Sequential()
network.add(layers.Dense(500, activation='relu'))
network.add(layers.Dense(200, activation='sigmoid'))
network.add(layers.Dense(9, activation='softmax'))

network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

network.fit(ktrainfeat, ktl, epochs=15, batch_size=100)

Epoch 1/15
619/619 [==============================] - 2s 2ms/step - loss: 0.7540 - accuracy: 0.7284
Epoch 2/15
619/619 [==============================] - 2s 2ms/step - loss: 0.5318 - accuracy: 0.7908
Epoch 3/15
619/619 [==============================] - 2s 2ms/step - loss: 0.4910 - accuracy: 0.8054
Epoch 4/15
619/619 [==============================] - 1s 2ms/step - loss: 0.4667 - accuracy: 0.8135
Epoch 5/15
619/619 [==============================] - 1s 2ms/step - loss: 0.4422 - accuracy: 0.8242
Epoch 6/15
619/619 [==============================] - 1s 2ms/step - loss: 0.4238 - accuracy: 0.8301
Epoch 7/15
619/619 [==============================] - 1s 2ms/step - loss: 0.3997 - accuracy: 0.8419
Epoch 8/15
619/619 [==============================] - 2s 2ms/step - loss: 0.3839 - accuracy: 0.8478
Epoch 9/15
619/619 [==============================] - 1s 2ms/step - loss: 0.3665 - accuracy: 0.8534
Epoch 10/15
619/619 [==============================] - 1s 2ms/step - loss: 0.3494 - accuracy: 0.8614

In [89]:
predictions = network.predict(kaggletest)
net = pd.DataFrame(predictions)
net.index += 1 
net.index.name = 'id'
net.columns=['Class_1','Class_2','Class_3','Class_4','Class_5','Class_6','Class_7','Class_8','Class_9']

In [90]:
net

Class_1       Class_2  ...       Class_8       Class_9
id                                  ...                            
1       3.696331e-05  7.341960e-01  ...  3.225402e-07  1.050146e-06
2       5.933168e-05  5.199404e-05  ...  9.721024e-02  3.131813e-04
3       2.366898e-11  1.438781e-08  ...  3.404164e-06  1.421668e-07
4       1.793807e-07  8.985019e-01  ...  2.413338e-07  8.204160e-06
5       6.223351e-02  3.559636e-06  ...  2.575807e-03  9.335529e-01
...              ...           ...  ...           ...           ...
144364  6.271170e-03  4.441265e-06  ...  3.840471e-03  5.590571e-04
144365  8.932214e-04  2.163854e-01  ...  1.839120e-06  8.079296e-06
144366  7.517587e-11  6.052749e-01  ...  6.141050e-07  1.635235e-06
144367  8.739545e-07  9.188178e-01  ...  6.567232e-07  1.593081e-06
144368  3.669070e-07  7.408299e-01  ...  2.069186e-06  2.218814e-06

[144368 rows x 9 columns]

In [119]:
boost4.to_csv('boostresults.csv')

In [120]:
net.to_csv('networkresults.csv')

#Conclusion



**Results**

XGBoost: public score = 5.89733 ||||| private score = 5.91635

Neural Network: public score = 0.54523 ||||| private score = 0.54644

**The images of my results are in the projects folder of my github repository.**

So my results were not exactly what I expected. My XGBoost classifier, which I thought would perform better, totally bombed. However, my neural network classifier performed quite well. I am not sure if I made a mistake somewhere, or if XGBoost really just did not work well. Regardless I am happy with my neural network score, 0.54 seems quite good!